Preview: storage & boot API for Image manager

nvlsianpu opened the pull request:
Low-level modules modules for DFU support:
- module for interact with mcuboot.
- module for storage the image.

# related PRs
 - DT flash write alignment property: https://github.com/zephyrproject-rtos/zephyr/pull/860

Signed-off-by: Andrzej Puzdrowski andrzej.puzdrowski@nordicsemi.no

nvlsianpu committed:
DFU: add module for interact with mcuboot

read, update status
trigger flashing
erase image bank
Module is intended to be use by a higher-level
image management protocol module.

Signed-off-by: Fabio Utzig <utzig@apache.org>
Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu committed:
DFU: add module for store the image

Store image in sequence of certain blocks.
Module is intended to be use by a higher-level
image management protocol module

Signed-off-by: Michael Scott <michael.scott@linaro.org>
Signed-off-by: Marti Bolivar <marti.bolivar@linaro.org>
Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

d3zd3z commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,135 @@
+ */
+
+#define TRAILER_IMAGE_MAGIC_SIZE (4 * sizeof(u32_t))
+#define TRAILER_SWAP_STATUS_SIZE (128 * FLASH_MIN_WRITE_SIZE * 3)
```
This looks like it might be out of date with the current mcuboot code. We've gotten rid of the need for update software to need to know the status size. The image magic has been moved to the end, and is the only thing that needs to be written (possibly with the image_ok already set).

d3zd3z commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,135 @@
+	copy_done_offset = boot_trailer_copy_done(FLASH_BANK1_OFFSET);
+	image_ok_offset = boot_trailer_image_ok(FLASH_BANK1_OFFSET);
+	memset(copy_done, TRAILER_PADDING, sizeof(copy_done));
+	memset(image_ok, TRAILER_PADDING, sizeof(image_ok));
```
It shouldn't be necessary to write values that are already 0xFF.  Also, in the new code, these are defined to always be 8 bytes long, so that this code doesn't need to know the flash write alignment.

d3zd3z commented on subsys/dfu/boot/targets/96b_nitrogen.h:
```diff
@@ -0,0 +1,20 @@
+
+#define FLASH_DRIVER_NAME		CONFIG_SOC_FLASH_NRF5_DEV_NAME
+#define FLASH_ALIGN			4
+#define FLASH_AREA_IMAGE_0_OFFSET	0x08000
```
These values should be coming from the device tree. We should add partition tables to the dtsi files for the targets we want, and not add the defines here. k64f should already have one as an example.

d3zd3z replied:
The only define that should be needed from these files is the FLASH_DRIVER_NAME, and I think we should be figuring out how to get the DT code to generate the define based on the chosen node.

nvlsianpu replied:
Do you mean to move these defines to /home/anpu/codedev/zephyr/boards/arm/f96b_nitrogen.h/board.h for instance?

d3zd3z replied:
To confirm, we need to solve the driver name issue. I think a future solution will be to have this defined entirely by the device tree (and not even using a Kconfig option). But, it seems clumsy to add a bunch of target-specific headers that each contain a single #define of a device name, but I'm not sure how else to handle it until the other aspect is solved. @agross-linaro thoughts?

d3zd3z commented on subsys/dfu/img_util/flash_block.c:
```diff
@@ -0,0 +1,137 @@
+static bool flash_verify(struct device *dev, off_t offset,
+			 u8_t *data, int len)
+{
+	int temp;
```
Int seems like an unusual type for this. You probably want something a bit bigger than calling through the flash API for every 4 bytes. If anything, how about making it an explicit 32-bit value, or an array of uint8 values?

d3zd3z commented on subsys/dfu/img_util/flash_block.c:
```diff
@@ -0,0 +1,137 @@
+
+	if (finished && cb->buf_bytes > 0) {
+		/* pad the rest of cb->buf and write it out */
+		memset(cb->buf + cb->buf_bytes, 0,
```
Padding on flash should be with 0xFF, so that it seems unwritten.

d3zd3z commented on subsys/dfu/boot/Makefile:
```diff
@@ -0,0 +1,2 @@
+obj-$(CONFIG_BOOT_IMG_MANAGER) += mcuboot.o
+ccflags-$(CONFIG_BOOT_IMG_MANAGER) += -I${ZEPHYR_BASE}/subsys/dfu
```
I believe 'ccflags' are the flags for the C compiler. C++ flags would be cppflags.

d3zd3z reviewed (changes requested):
The main thing I think that this can be simplified a bit with some changes that have gone into mcuboot.

Presumably, there is more code coming that implements the DFU itself?

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,135 @@
+/*
+ * Copyright (c) 2016-2017 Linaro Limited
```
I will add   **Copyright (c) 2017 Nordic Semiconductor ASA** everywhere.

mbolivar commented on subsys/dfu/img_util/flash_block.h:
```diff
@@ -0,0 +1,61 @@
+ * @param cb context
+ * @param data data to write
+ * @param len Number of bytes to write
+ * @param finished when true this forces any buffered
```
Since this is about flushing a buffer, maybe rename this argument "flush" instead for submission upstream?

I also think that the documentation here should add some extra discussion about what it means to flush these bytes. 

The routine will pad the buffer out with 0xff up to the next CONFIG_IMG_BLOCK_BUF_SIZE boundary before writing, and not all flash hardware allows writing 0xff more than once. So the entire flash device up to the next multiple of CONFIG_IMG_BLOCK_BUF_SIZE is no longer safe to write until the next erase. This may be more bytes than what the user expects, so it seems like a good idea to make a note of that in the documentation.

mbolivar commented on subsys/dfu/img_util/flash_block.c:
```diff
@@ -0,0 +1,137 @@
+			return rc;
+		}
+
+		if (!rc &&
```
The `!rc &&` can be deleted, since you've already checked rc above.

mbolivar commented on subsys/dfu/img_util/flash_block.c:
```diff
@@ -0,0 +1,137 @@
+			return rc;
+		}
+
+		if (!rc &&
```
Same here, no need to check rc again.

mbolivar commented on subsys/dfu/Kconfig:
```diff
@@ -0,0 +1,24 @@
+# Kconfig - DFU support configuration options
```
It's either that or "lib". I expect this will grow over time (especially once we start querying mcuboot capabilities etc. at runtime) into something that looks more like a subsystem than a library, so I agree.

nashif replied:
subsys is fine IMO

mbolivar reviewed (changes requested):
Hi Andrzej,

Thank you so much for picking this code up and working on getting it upstreamed, and thanks a lot for the preview of your work in progress!

I've left some review comments. Overall I think this is going in a great direction, and thanks again for taking this task on.

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,254 @@
+#define BOOT_FLAG_IMAGE_OK 0
+#define BOOT_FLAG_COPY_DONE 1
+
+#define COPY_DONE_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE -\
```
If to be onset I was thinking about using [image trailer struct](https://github.com/runtimeco/mcuboot/blob/master/boot/bootutil/include/bootutil/bootutil.h#L66) to do so.

mbolivar replied:
I don't think it's a good idea to use that file, for reasons I explained when the pull request which added it was submitted to mcuboot: https://github.com/runtimeco/mcuboot/pull/76#issuecomment-312341305

d3zd3z replied:
You might check the generated code, but at least in the past __packed had a significant impact on the generated code. It might not matter in this case, since you're only accessing a byte field, but in the general case, it does revert to reading/writing all fields with byte accesses.

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,254 @@
+				  BOOT_MAX_ALIGN)
+#define MAGIC_OFFS(bank_offs) (bank_offs + FLASH_BANK_SIZE - BOOT_MAGIC_SZ)
+
+const u32_t boot_img_magic[4] = {
```
You have right, although it is similar in mcuboot sources. Maybe missed I something?

mbolivar replied:
Sounds like a bug in mcuboot :).

d3zd3z replied:
Not necessarily a bug. We could just describe the magic in terms of 32-bit values that must match the endianness of the target. Realistically, I'm not aware of a big-endian CPU supported by Zephyr.

nvlsianpu replied:
I would keep as is here and in mcuboot - For both projects this is internal-only entity and its endianes is determinates by the target used.

mbolivar replied:
> Realistically, I'm not aware of a big-endian CPU supported by Zephyr.

arc and xtensa both support big-endian configurations.

mbolivar replied:
OK, I'm wrong: the mcuboot design document specifies that these bytes are written in host byte order, not little-endian order, so this code is OK.

d3zd3z commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,254 @@
+	return flag_val;
+}
+
+static u8_t boot_copy_done_get(u32_t bank_offs)
```
Do we need to have this function at all? The copy-done is pretty much internal to the bootloader, and shouldn't need to be read here.

d3zd3z commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,254 @@
+
+	for (i = 0; i < 4; i++) {
+		if (magic[i] == 0xffffffff) {
+			return BOOT_MAGIC_UNSET;
```
Yes, there are definitely 3 states: Unset, meaning all 0xFFFFFFFF, Set, exactly matching the magic, and Invalid. It is important, because if it is invalid, it cannot be written to, and at least the sector will have to be erased.

nvlsianpu replied:
Yes It is bug. [mynewt](https://github.com/apache/mynewt-core/blob/master/boot/bootutil/src/bootutil_misc.c#L152) as well has it.

d3zd3z commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,254 @@
+{
+	int rc;
+
+	switch (boot_magic_state_get(FLASH_BANK1_OFFSET)) {
```
I don't quite understand the logic, or for that matter why there is so much code here.  Honestly, the only thing all of this code needs to do is, after placing the image in place, write the magic at the end, and write the "image_ok" marker at the end. There doesn't need to be any logic, or even reading of the device, since the partition should be erased by the process of writing the image.

nvlsianpu replied:
This is the logic straight from mcuboot sources, and it is defensive programming (which I realy like).  As you recommended I will reduce code here.

d3zd3z commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,254 @@
+		return -EFAULT;
+	}
+
+	if (boot_copy_done_get(FLASH_BANK0_OFFSET) == BOOT_FLAG_UNSET) {
```
This is kind of pointless. This code will never be running if the copy isn't completed. Really, none of this code makes sense. It is fine to just write the image_ok and magic value, and be done. There isn't anything the code could do if something were to go wrong anyway. All of this logic just makes it confusing, and isn't going to help anything.

d3zd3z commented on subsys/dfu/boot/mcuboot.h:
```diff
@@ -0,0 +1,41 @@
+
+/**
+ * @brief Marks the image in slot 0 as confirmed. The system will continue booting
+ * into the image in slot 0 until told to boot from a different slot.
```
Not quite accurate. There are two ways of doing this:
1. One-shot upgrade. Write the new image in slot 1, add image magic, and set image_ok. The bootloader will move to slot 0, and boot it from then on.
2. Trial upgrade. Write the image in slot 1, and add image magic. Bootloader will swap images, and boot slot 0. If image works, it should set image_ok in slot 0.

However, it needs to be clear that this function really isn't part of the DFU or upgrade, but needs to be done on every boot.

d3zd3z replied:
But, we need to be able to mark this for both slots, to be able to perform either upgrade type.

d3zd3z commented on subsys/dfu/boot/mcuboot.h:
```diff
@@ -0,0 +1,41 @@
+ *   1=run image forever.
+ * @return 0 on success, negative errno code on fail.
+ */
+int boot_img_pending_set(int permanent);
```
This doesn't quite make sense. The operation we should have is to mark slot 1 for upgrade, with a permanent flag. It seems like there might be a better name for that, "boot_request_upgrade" perhaps?

d3zd3z commented on subsys/dfu/boot/targets/nrf52840_pca10056.h:
```diff
@@ -0,0 +1,16 @@
+
+#define FLASH_DRIVER_NAME		CONFIG_SOC_FLASH_NRF5_DEV_NAME
+#define FLASH_ALIGN			4
+#define FLASH_AREA_IMAGE_0_OFFSET	0x08000
```
Same here. That you aren't getting redefinition warnings suggests that this device isn't in the device tree. The partitions should be added there, and not with these defines.

d3zd3z commented on subsys/dfu/img_util/flash_block.c:
```diff
@@ -0,0 +1,137 @@
+	while ((len - processed) >
+	       (CONFIG_IMG_BLOCK_BUF_SIZE - cb->buf_bytes)) {
+		memcpy(cb->buf + cb->buf_bytes, data + processed,
+		       (CONFIG_IMG_BLOCK_BUF_SIZE - cb->buf_bytes));
```
Why can't we just write directly from 'data'? Or is this buffer somehow coming in misaligned (some drivers won't handle that).

d3zd3z reviewed (changes requested):
In addition to the code comments, the commit text needs to be expanded quite a bit. Think of the commit text as where you justify "why" these commits should be brought into the tree. They are also important for someone reading the code later to understand why a given piece of code is there. What is the goal, etc. I would expect at least a few sentences to describe each of these patches, and they should be complete sentences, not just fragments.

d3zd3z commented on dts/arm/nordic/nrf52840.dtsi:
```diff
@@ -10,6 +10,10 @@
 
 	flash0: flash {
 		reg = <0x00000000 DT_FLASH_SIZE>;
+		prop: flash-properies {
```
Why is this a full node, rather than just putting the alignment as a property within the flash itself?

Let's make the name itself more descriptive than just "align", since it is specifically the write alignment. I don't really like the term "align", since it only implies what boundary the writes must happen on, but not that the writes have to be done in units of that size. How about something like "write-block-size".

Unsure if this should be "write-block-size", or "zephyr,write-block-size", hopefully @agross-linaro has thoughts on this.

nvlsianpu commented:
@d3zd3z, @agross-linaro I've opened separate PR for addition of flash alignment description. https://github.com/zephyrproject-rtos/zephyr/pull/860

mbolivar commented on subsys/dfu/img_util/flash_img.c:
```diff
@@ -0,0 +1,137 @@
+ */
+
+#define SYS_LOG_DOMAIN "fota/flash_block"
+#define SYS_LOG_LEVEL SYS_LOG_LEVEL_ERROR
```
Sorry I was unclear in my earlier review.

What Zephyr usually does is to define a CONFIG_SYS_LOG_FOO_LEVEL (see for example CONFIG_SYS_LOG_CLOCK_CONTROL_LEVEL, CONFIG_SYS_LOG_SENSOR_LEVEL, CONFIG_SYS_LOG_GPIO_LEVEL) for each "FOO" which wants to log something.

To do the same thing, you can make a new Kconfig option named SYS_LOG_MCUBOOT_LEVEL, which defaults to the error value, then write `#define SYS_LOG_LEVEL CONFIG_SYS_LOG_MCUBOOT_LEVEL` here and in any other files in this subsystem that need to log.

That will let the users change the log level without editing all the sources, just by making changes to their project configuration file.

mbolivar commented on subsys/dfu/boot/mcuboot.h:
```diff
@@ -0,0 +1,43 @@
+
+/**
+ * @brief Marks the image in slot 1 as pending. On the next reboot, the system
+ * will perform a one-time boot of the slot 1 image.
```
This won't be a "one-time" boot if "permanent" is set, will it?

mbolivar commented on subsys/dfu/boot/mcuboot.h:
```diff
@@ -0,0 +1,43 @@
+ *
+ * @return 0 on success, negative errno code on fail.
+ */
+int boot_img_confirmed_write(void);
```
Just a nit: "boot_img_confirmed_write" is "subject object verb" order, and "boot_request_upgrade" is "subject verb object" order.

Since "subject verb object" is the usual order in English, how about renaming this to "boot_write_img_confirmed", and similarly using "boot_erase_img_bank" below? To be clear, I'd be fine with doing that as a cleanup patch that gets sent later on.

mbolivar commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,263 @@
+
+	rc = flash_erase(flash_dev, bank_offset, FLASH_BANK_SIZE);
+	if (rc) {
+		return rc;
```
This is tricky: you still want to try to write-protect the flash here instead of returning right away.

mbolivar replied:
The API says: "on some flash components, the write protection is automatically turned on again by the device after the completion of each write or erase calls"

The operative word here is "completion", not "successful completion."

nvlsianpu replied:
Do you mean that flash should be protected back even in case erase operation had failed?

mbolivar commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,263 @@
+		return BOOT_MAGIC_GOOD;
+	}
+
+	for (i = 0; i < 4; i++) {
```
i < ARRAY_SIZE(boot_img_magic)

mbolivar commented on dts/common/yaml/flash.yaml:
```diff
@@ -0,0 +1,26 @@
+base_label: FLASH
+use-property-label: yes
+
+...
```
Looks like this file is missing a newline at the end.

mbolivar commented on subsys/dfu/img_util/flash_img.h:
```diff
@@ -0,0 +1,63 @@
+struct flash_img_context {
+	u8_t buf[CONFIG_IMG_BLOCK_BUF_SIZE];
+	struct device *dev;
+	u32_t bytes_written;
```
This is still a u32_t instead of a size_t -- can you explain why that is better, since the flash API counts bytes in terms of size_t?

mbolivar commented on subsys/dfu/img_util/flash_img.h:
```diff
@@ -0,0 +1,63 @@
+ * memory in single blocks. Will store remainder between calls.
+ *
+ * A final call to this function with flush set to true
+ * will write out the remaining block buffer to flash. As One block is always
```
Thanks for fixing this! Note that "One" should be lower-cased.

(You might want to say something like "Since flash is written to in blocks, the contents of flash from the last byte written up to the next multiple of CONFIG_IMG_BLOCK_BUF_SIZE is padded with 0xff".)

mbolivar commented on subsys/dfu/img_util/flash_img.c:
```diff
@@ -0,0 +1,137 @@
+		}
+
+		if (memcmp(data, &temp, size)) {
+			SYS_LOG_ERR("offset=0x%08x VERIFY FAIL\n\r"
```
Awesome, thanks for adding expected/actual.

However, adding the \n\r will confuse sys log parsers (including humans), which expect each log message to be on one line (and may or may not use \n -- see SYS_LOG_NO_NEWLINE). Can you delete the \n\r?

mbolivar reviewed (changes requested):
Thanks for the update. Things are looking much cleaner now. I've left a (smaller) number of comments on this version; I'll defer to others on the DT changes.

rsalveti commented on subsys/Makefile:
```diff
@@ -8,3 +8,4 @@ obj-$(CONFIG_DISK_ACCESS) += disk/
 obj-$(CONFIG_CPLUSPLUS) += cpp/
 obj-y += logging/
 obj-y += debug/
+obj-$(CONFIG_BOOT_IMG_MANAGER) += dfu/
```
Should we have a DFU or BOOT_DFU option to match the directory name?

rsalveti commented on subsys/dfu/Kconfig:
```diff
@@ -0,0 +1,24 @@
+	prompt "Image writer buffer size"
+	default 512
+	help
+	  Size (in Bytes) of buffer for image writer. Must be a multiple of the access alignment required by used flash driver.
```
Break into 2 lines.

nashif commented on subsys/Makefile:
```diff
@@ -8,3 +8,4 @@ obj-$(CONFIG_DISK_ACCESS) += disk/
 obj-$(CONFIG_CPLUSPLUS) += cpp/
 obj-y += logging/
 obj-y += debug/
+obj-$(CONFIG_MCUBOOT_IMG_MANAGER) += dfu/
```
why are we making a subsys like dfu depend on a signle bootloader implementation like MCUBOOT? We need to abstract this and have the MCU_IMG_MANAGER be an implementation, i.e. one out of possible many.

nvlsianpu replied:
I changed this. I added menuconfig at top level:
`      DFU options  --->    `
and under it I putted follow configs:
`  [*] DFU image manager                                            `
`            Image manager (Image manager for mcuboot)  --->             `
`      (512) Image writer buffer size                                    `
`      (0) Mcuboot image manager Log level  `

nashif commented on subsys/dfu/boot/Makefile:
```diff
@@ -0,0 +1,2 @@
+obj-$(CONFIG_MCUBOOT_IMG_MANAGER) += mcuboot.o
```
You are using CONFIG_MCUBOOT_IMG_MANAGER before it was defined in Kconfig.  Please define Kconfig variable in the same commit they are being used.

nvlsianpu replied:
I did as you recommend.

nashif reviewed (changes requested):
Can we add a sample or a test that exercises this code addition?

nvlsianpu commented:
@nashif I apply requested changes.

Regards exercising the code
- sample code: I want to wait for DFU protocol support contribution for building sample DFU application. We will contribute DFU protocol support as one of next step.

- tests: I can add test for verification whether added APIs modifies flash memory according to expectations. Will you be happy with such  tests?

nashif commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,263 @@
+
+#define FLASH_MIN_WRITE_SIZE FLASH_ALIGN
+#define FLASH_BANK0_OFFSET FLASH_AREA_IMAGE_0_OFFSET
+#define FLASH_BANK_SIZE FLASH_AREA_IMAGE_0_SIZE
```
where is FLASH_AREA_IMAGE_0_SIZE defined? same for FLASH_AREA_IMAGE_0_OFFSET? Is this coming from the board or Soc? Does seem to be defined anywhere in Zephyr

nashif replied:
ok, I think it is worth adding a comment, not all architectures and boards in Zephyr have DTS support. It will help understand where this is coming from.

nvlsianpu commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,263 @@
+
+	rc = flash_erase(flash_dev, bank_offset, FLASH_BANK_SIZE);
+	if (rc) {
+		return rc;
```
@kl-cruz

nvlsianpu committed:
tests: subsys: Add tests for new DFU modules

Added tests for the mcuboot interface module and the image
storage module.
Both tests are checking flash content produced by modules verified.

Signed-off-by: Andrzej Puzdrowski <andrzej.puzdrowski@nordicsemi.no>

nvlsianpu commented:
@nashif @mbolivar @d3zd3z @rsalveti I added tests for new modules. Tests verifies flash content which is writing by modules.

mbolivar commented on subsys/dfu/boot/mcuboot.c:
```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright (c) 2017 Nordic Semiconductor ASA
+ * Copyright (c) 2016-2017 Linaro Limited
```
LGTM

mbolivar commented on subsys/dfu/boot/mcuboot.h:
```diff
@@ -0,0 +1,41 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef __MCUBOOT_H__
```
Can you give more specifics on this? I could imagine using this code from applications (as you probably know, we do already use code like this from applications).

nvlsianpu replied:
Now I see a case (thanks to dfu protocols analyses I've done) - this will be moved to include directory.

carlescufi commented:
@nashif @cvinayak could you re-review this at your convenience?

carlescufi commented on tests/subsys/dfu/mcuboot/src/main.c:
```diff
@@ -0,0 +1,138 @@
+#include <dfu/mcuboot.h>
+
+/* FLASH_AREA_IMAGE_XX_YY values used below are auto-generated thanks to DT */
+#define FLASH_BANK1_SIZE FLASH_AREA_IMAGE_1_SIZE
```
Why redefine those here? Wouldn't it be easier to just use the DT macros?
